src/storage/src/source/mysql/replication.rs

```diff
@@ -462,15 +470,168 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                         upper_cap_set.downgrade(&*new_upper);
 
                         rewinds.retain(|_, (_, req)| {
-                            !PartialOrder::less_than(&req.snapshot_upper, &new_upper)
+                            // We need to retain the rewind requests whose snapshot_upper has
+                            // at least one timestamp such that new_upper is less than that
+                            // timestamp
+                            req.snapshot_upper.iter().any(|ts| new_upper.less_than(ts))
                         });
                         trace!(%id, "timely-{worker_id} pending rewinds \
                                      after filtering: {rewinds:?}");
                     }
+                    // From the MySQL docs: 'A Query_event is created for each query that modifies
+                    // the database, unless the query is logged row-based.' This means that we can
+                    // expect any DDL changes to be represented as QueryEvents, which we must parse
+                    // to figure out if any of the tables we care about have been affected.
+                    Some(EventData::QueryEvent(event)) => {
```

Comment:
From a maintainability perspective, `render` is already a rather long function, spanning multiple hundred lines. This change adds some more lines and nesting depth to it. If possible, you could consider extracting some functions. (just a suggestion, optional!)